### DISCUSSION TITLE
Character/player issues

### DISCUSSION THREAD

**Discussion by RandomEC** [evennia/evennia#2719] [Evennia Questions] [answered]
So, I've been having an issue with my resetting code. Unlike most of the problems I've had to date, I've had difficulty replicating this one, making it difficult to figure out the cause. Basically, the code that I use to reset my areas in my mud follows standard Envy-type protocol, when it goes to reset, it checks if there is a player in the area. If there isn't, it resets, if there is, it increments a variable stored on the resetting script. If the variable is at 2 when it goes to reset, it resets regardless of whether there is a player in the area. By and large it works fine ... except when it doesn't. Recently (before I knew I had a problem), I came up with some code to streamline figuring out if there was a player in the area. When I tried to plug this in for my old means of doing the same thing, I noticed the problem. I thought this was the issue, but when I reverted to the old reset mechanism, I noticed an issue with that this morning as well. So my suspicion is that it is something to do with player behavior, but I can't figure out anything obvious. I should also note that I have had instances where the new player-detection code works absolutely fine in other contexts.

I've tried somewhat to debug this, but obviously had limited success. I guess my question is whether there is some odd behavior of characters that I'm not aware of. Absolutely any advice is appreciated.

Reset Script with old player-detection code
```
from evennia.utils import search

class ResetScript(DefaultScript):

    def at_script_creation(self):
        self.key = "reset_script"
        self.interval = settings.EMPTY_AREA_RESET_TIME
        self.desc = "Handles resets for Echoes of Creation"
        self.persistent = True

        # The below is a list of areas in the MUD, paired with how many cycles
        # have gone by since resetting.

        # Maybe make a command to add an area to this list.

        self.db.area_list = {
            "smurf village": 0,
            "graveyard": 0,
            "haon dor": 0,
            "dwarven daycare": 0,
            "training tower": 0,
            "the circus": 0,
            "the library": 0,
            "edens grove": 0,
            "crystalmir lake": 0,
            "the rats' lair": 0
        }


    def check_for_player(self, area_name):
        """
        This function checks to see if there are players in the area in
        question.
        """

        # Get all the objects in the area
        all_area_objects = search.search_tag(area_name, category="area name")
        # Iterate through all objects to see if there are any players in rooms
        # in the area
        for object in all_area_objects:
            contents = object.contents
            for item in contents:
                # If there is a player, return true.
                if item.account:
                    return True
        return False

    def at_repeat(self):

        for area in self.db.area_list:
            # Reset if there are no players in the area, or if counter is at 2.
            if not self.check_for_player(area) or self.db.area_list[area] == 2:
                objects_to_reset = search.search_tag(area, category="area name")
                if objects_to_reset:
                    for object in objects_to_reset:
                        object.at_reset()
                # Since you reset, reset the timer on the area.
                self.db.area_list[area] = 0
            else:
                # Iterate up the timer toward the max of two.
                self.db.area_list[area] += 1
```

New player-detection code
```
from evennia.utils import search

def player_in_area(area_name):
    """
    This function checks whether there is a player in an area
    or not.
    """
    players = search.search_tag("player")
    
    if players:
        for player in players:
            if player.location:
                if area_name in player.location.tags.all():
                    return True
    
    return False
```

Thanks in advance!

**Comment by InspectorCaracal** [comment, marked as answer]
What does "except when it doesn't" mean? >.> Does it fail to reset when it should, or does it reset when it shouldn't?

**Reply by RandomEC** [reply]
Good point. It fails to reset when it should.

**Reply by InspectorCaracal** [reply]
Yeah, I don't see anything obvious in there that could be returning a false _positive_ for players being present, which is the only way I can see it missing a reset or an increment.

I'd have the script log everything on its reset attempt to figure out where the problem is happening. Log which area it's checking, then which players it finds, what those players locations are, and whether it's resetting the area. Then you can look over the log for inconsistencies or things you aren't expecting to see, which'll help track down the problem.

**Reply by InspectorCaracal** [reply]
As a side note, if you're looping through all the players anyway, I'd do that first and collect a list of all the zones they're in so you only need to iterate through the players once; then you can just check if each area is in that list instead of doing a player search each time.

**Reply by RandomEC** [reply]
Heh. So I'll have to figure out logging, then. I've made it this far with just using player.msg to check stuff. Oh well, I knew I'd have to learn it eventually.

I was thinking that the player_in_area function was erroring, which was causing the whole thing to stop in its tracks, not necessarily a false positive. This is particularly the case as resetting worked before and after I slotted that in. But I can't see anything in the player_in_area function that should be wrong. I've protected against getting a None list for players (if players) and a player being at None for some reason (if player.location). I don't know what else could be wrong with it. It's not a complicated function.

As to what's going wrong with it as currently configured (with check_for_player), I've no idea. Since player_in_area touches players EVERY time it fires, and check_for_player only interacts with players when it finds one, I figured that the players somehow were the problem, as the first one screws up every time, and the second one is intermittent. I just don't know what the problem could be. Also, incidentally, the area that didn't reset with the current ("working") code was an area that had a non-active player in it.

**Reply by RandomEC** [reply]
Fair point, though I use that function for a couple of different things. I was privileging portability over efficiency.

**Reply by InspectorCaracal** [reply]
> the area that didn't reset with the current ("working") code was an area that had a non-active player in it.

Oh, well, that's an important piece of information, because that's why it's failing! There's a player in it, so when it checks if there's a player in it... there is! I (wrongly) assumed players were being moved out of the IC map when logging out so I didn't think of that.

If you want to check if a player is active, you can use `player.has_account`

**Reply by RandomEC** [reply]
Well, but the way the code works, if there is a player in the area, it should drop into the "else" in the at_repeat for loop, and bump up by one. After doing that twice, it should reset anyway because it hits on the self.db.area_list[area] == 2. It's worth mentioning that should take about 15 minutes (5 min per rep). It didn't reset overnight.

**Reply by InspectorCaracal** [reply]
Did you check the area_list attribute? Because since it only checks if equal 2, rather than equal or greater to 2, if it incremented to over 2 due to the other method not working, it'll never trigger.

**Reply by RandomEC** [reply]
Yeah, I caught that this morning when I posted this. It's at home, so I can't easily check from here. There's still a problem in that case, though, as it is for some reason skipping over 2.

**Reply by RandomEC** [reply]
InspectorCaracal, thanks for all the help, by the way. I'm certainly going to build in an active player check, and if it hopped over 2 by some freak thing, just doing >= 2 should fix it.

I feel like when I tested player_in_area, though, I was having trouble getting it to recognize a player in the area. But it's possible that I may have tested it within the scheme of the at_repeat (which I "knew" worked, at the time), so it's possible it was erroring for the > 2 reason then, too.

Well, we'll see tonight! Thanks again, as usual, for all the help! Subject to cleaning up some minor bugs, I do have a functioning, playable mud now, even if it IS only for five levels. Ah well, we're getting there. :)))

**Reply by InspectorCaracal** [reply]
No problem!

I can think of some reasons why `player_in_area` might be thinking a zone was empty of players (such as, are you making sure all players have the `player` tag?) but I'm at a loss for why it might return a false positive or break enough to never finish the reset check. Good luck debugging 👍

**Reply by RandomEC** [reply]
Agreed - I had thought of the same thing, in fact I know a player or two that does not have the tag, but I came to the same conclusion - that it shouldn't be a problem, as that would give you too MUCH activity, not too little. All new players get it at creation, so it should not be a problem going forward.

And thanks again. Generally speaking I've been able to figure out most things by myself, but this one had me stumped ...
